pipelining/test_pipelining.py

```diff
@@ -0,0 +1,228 @@
+    srv = None
+    rtype = server['response']
+    ka = server['keep_alive']
+    if rtype == 'static':
+        content = fill_template(server['response_content'])
+        srv = DeproxyKeepaliveServer(port=port,
+                                     response=content,
+                                     keep_alive=ka)
+    else:
+        raise Exception("Invalid response type: %s" % str(rtype))
+    tester.deproxy_manager.add_server(srv)
+    return srv
+
+tester.register_backend('deproxy_ka', build_deproxy_keepalive)
+
+class PipeliningTest(tester.TempestaTest):
+
+    backends = [
+        {
+            'id' : 'deproxy',
+            'type' : 'deproxy',
+            'port' : '8000',
+            'response' : 'static',
+            'response_content' : """HTTP/1.1 200 OK
+Content-Length: 0
+Connection: keep-alive
+
+"""
+        },
+        {
+            'id' : 'deproxy_ka',
+            'type' : 'deproxy_ka',
+            'keep_alive' : 4,
+            'port' : '8000',
+            'response' : 'static',
+            'response_content' : """HTTP/1.1 200 OK
+Content-Length: 0
+Connection: keep-alive
+
+"""
+        },
+    ]
+
+    tempesta = {
+        'config' : """
+cache 0;
+listen 80;
+nonidempotent GET prefix "/";
+
+srv_group default {
+    server ${general_ip}:8000;
+}
+
+vhost default {
+    proxy_pass default;
+}
+""",
+    }
+
+    clients = [
+        {
+            'id' : 'deproxy',
+            'type' : 'deproxy',
+            'addr' : "${tempesta_ip}",
+            'port' : '80'
+        },
+    ]
+
+    def test_pipelined(self):
+        request = "GET / HTTP/1.1\r\n" \
+                  "Host: localhost\r\n" \
+                  "\r\n" \
+                  "GET /path HTTP/1.1\r\n" \
+                  "Host: localhost\r\n" \
+                  "\r\n" \
+                  "GET /uri HTTP/1.1\r\n" \
+                  "Host: localhost\r\n" \
+                  "\r\n" \
+                  "GET /app HTTP/1.1\r\n" \
+                  "Host: localhost\r\n" \
+                  "\r\n"
+        deproxy_srv = self.get_server('deproxy')
+        deproxy_srv.start()
+        self.assertEqual(0, len(deproxy_srv.requests))
+        self.start_tempesta()
+        self.assertTrue(deproxy_srv.wait_for_connections(timeout=1))
+        deproxy_cl = self.get_client('deproxy')
+        deproxy_cl.start()
+        deproxy_cl.make_request(request)
+        resp = deproxy_cl.wait_for_response(timeout=5)
+        self.assertTrue(resp, "Response not received")
+        self.assertEqual(4, len(deproxy_cl.responses))
+        self.assertEqual(4, len(deproxy_srv.requests))
+        for i in range(len(deproxy_srv.requests)):
+            tf_cfg.dbg(3, "Req %i: %s" % (i, deproxy_srv.requests[i].msg))
+
+        self.assertEqual(deproxy_srv.requests[0].uri, "/")
+        self.assertEqual(deproxy_srv.requests[1].uri, "/path")
+        self.assertEqual(deproxy_srv.requests[2].uri, "/uri")
+        self.assertEqual(deproxy_srv.requests[3].uri, "/app")
```

Comment:
Totally wrong test case. See my notice at line 89 near `nonidempotent GET prefix "/";`.

Client pipelines only idempotent request. Exact order of requests on server doesn't matter, and even more, the requests can be distributed among multiple server connections and servers depending on LB configuration.

The test case must check that the client receives responses in the correct order: response to `/` then to `/path`. then to `/uri` and finally to the `/app`. Ideal test case would mix up order of the responses returned by the server via it's connection. E.g. `/path` request was forwarded via server connection №3, `/app` - via connection №5; the server will hold on transmission over connection №3 until response on connection №5 is sent. In this case responses will be returned to Tempeste in some random order, while Tempesta has to restore the original order before forwarding to the client.

In more complex test case server connection errors should be simulated, both expected (like depoxy_ka in this test) and unexpected (error during forwarding request/response). Forwarding of non-idempotent requests to backend is a subject for distinct test case.

Reply:
Just forgot a condition for a test case flavour: there should be multiple clients, each must receive it's own request.

Reply:
created tast #47 47